staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -599,6 +607,34 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		s.NonLongRunningRequestWaitGroup.Wait()
 	}()
 
+	// wait for all in-flight watches to finish
+	activeWatchesDrainedCh := make(chan struct{})
+	go func() {
+		defer close(activeWatchesDrainedCh)
+
+		<-notAcceptingNewRequestCh.Signaled()
+
+		// Wait for all active watches to finish
+		grace := 10 * time.Second
+		activeBefore, activeAfter, err := s.WatchRequestWaitGroup.Wait(func(count int) (utilwaitgroup.RateLimiter, context.Context, context.CancelFunc) {
+			ctx, cancel := context.WithTimeout(context.Background(), grace)
+			// TODO: a) do we need to make this grace period configurable?
```

Comment:
In my opinion we definitely need it to be more configurable.
I expect values of 30-60s to be used in practice for larger clusters.

Reply:
do you want this in the same PR, or a followup PR?